llama/sampling_ext.cpp

```diff
@@ -21,33 +22,31 @@ struct llama_sampling_context *llama_sampling_cinit(struct llama_sampling_cparam
     sparams.penalize_nl = params->penalize_nl;
     sparams.seed = params->seed;
     sparams.grammar = params->grammar;
-    return llama_sampling_init(sparams);
+    return (llama_sampler*)gpt_sampler_init(model, sparams);
```

Comment:
Good call! I was going for minimally invasive, but given the size of the PR at this point, I think keeping to the spirit rather than the text is better.

Reply:
It looks like using `gpt_sampler *` directly here will cause some problems because it requires including `sampling.h` which in turn brings in c++ headers like `<string>`. I'm not a C programmer at heart, so in C++, I'd use a forward declaration to get around this, but that doesn't seem to work here, so I _think_ the best bet is to stick with the pure-c interface in `llama.h` and the "base struct" `llama_sampler *` with the casting in `sampling_ext.cpp`. Does that sound reasonable?

Reply:
I think a forward declaration is the right solution. I tried it quickly and it seemed to work, what problem did you run into?